File: packages/guards/src/time-zone-safety.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface TimeZoneSafetyOptions {
  action: 'block' | 'warn';
}

const TZ_PATTERNS: RegExp[] = [
  /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?!Z|[+-]\d{2}:?\d{2})\b/,
  /\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}(?:\.\d+)?(?!\s*(?:UTC|GMT|Z|[+-]\d{2}|[A-Z]{2,5}))\s/,
  /(?:UTC|GMT)\s.*(?:EST|PST|CST|MST|CET|IST|JST|KST)/,
  /(?:EST|PST|CST|MST)\s.*(?:UTC|GMT)/,
  /(?:at|by|before|after)\s+\d{1,2}:\d{2}\s*(?:am|pm|AM|PM)?(?!\s*(?:UTC|GMT|ET|PT|CT|MT|[A-Z]{2,5}))\s/,
  /new\s+Date\(\s*['"]?\d{4}-\d{2}-\d{2}['"]?\s*\)/,
  /toLocaleString\(\)\s.*toISOString\(\)/,
];

export function timeZoneSafety(options: TimeZoneSafetyOptions): Guard {
  return {
    name: 'time-zone-safety',
    version: '0.1.0',
    description: 'Detects timezone-related issues in text',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of TZ_PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'time-zone-safety',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/time_zone_safety.py
"""Detect timezone-related issues in text."""

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}"
        r"(?:\.\d+)?(?!Z|[+-]\d{2}:?\d{2})\b"
    ),
    re.compile(
        r"\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}"
        r"(?:\.\d+)?"
        r"(?!\s*(?:UTC|GMT|Z|[+-]\d{2}|[A-Z]{2,5}))\s"
    ),
    re.compile(
        r"(?:UTC|GMT)\s.*(?:EST|PST|CST|MST|CET|IST|JST|KST)"
    ),
    re.compile(r"(?:EST|PST|CST|MST)\s.*(?:UTC|GMT)"),
    re.compile(
        r"new\s+Date\(\s*['\"]?\d{4}-\d{2}-\d{2}"
        r"['\"]?\s*\)"
    ),
]


class _TimeZoneSafety:
    def __init__(
        self, *, action: str = "warn"
    ) -> None:
        self.name = "time-zone-safety"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0) if triggered else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="time-zone-safety",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Timezone issues detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def time_zone_safety(
    *, action: str = "warn"
) -> _TimeZoneSafety:
    return _TimeZoneSafety(action=action)
